### PR TITLE
Python code for BIT and Segment Tree

### PR DESCRIPTION
**Pull request by thatpythonguys**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

File: content/4_Gold/PURS.mdx [hunk 1]
```diff
@@ -314,14 +314,76 @@ public class MinSegmentTree {
...
+
+	def update(self, i, v) -> None:
+		i += self.n
+		self.tree[i] = v
+
+		while i > 1:
+			self.tree[i >> 1] = min(self.tree[i], self.tree[i ^ 1])
+			i >>= 1
+
+	def query(self, l, r) -> int:
```
Comment (SansPapyrus683) [inline content/4_Gold/PURS.mdx:344]: argument type annotations

File: content/4_Gold/PURS.mdx [hunk 2]
```diff
@@ -314,14 +314,76 @@ public class MinSegmentTree {
...
 
 </JavaSection>
+
+<PySection>
+```python
+import sys
+inf = float('inf')
+input = sys.stdin.readline
+
+# --- BEGIN SEGMENT TREE---
```
Comment (SansPapyrus683) [inline content/4_Gold/PURS.mdx:324]:
```suggestion
```

File: content/4_Gold/PURS.mdx [hunk 3]
```diff
@@ -314,14 +314,76 @@ public class MinSegmentTree {
...
+			if l & 1:
+				ret = min(ret, self.tree[l])
+				l += 1
+			if r & 1:
+				r -= 1
+				ret = min(ret, self.tree[r])
+			l >>= 1
+			r >>= 1
+		return ret
+# --- END SEGMENT TREE ---
```
Comment (SansPapyrus683) [inline content/4_Gold/PURS.mdx:358]:
```suggestion
```

File: content/4_Gold/PURS.mdx [hunk 4]
```diff
@@ -314,14 +314,76 @@ public class MinSegmentTree {
...
+
+
+def solution():
```
Comment (SansPapyrus683) [inline content/4_Gold/PURS.mdx:361]: could you take this out of the function?

File: content/4_Gold/PURS.mdx [hunk 5]
```diff
@@ -591,6 +709,59 @@ int main() {
...
 }
 ```
+</CPPSection>
+<PySection>
+```python
+import sys
+input = sys.stdin.readline
+inf = float('inf')
+
+# --- BEGIN FENWICK TREE ---
```
Comment (SansPapyrus683) [inline content/4_Gold/PURS.mdx:719]:
```suggestion
```

File: content/4_Gold/PURS.mdx [hunk 6]
```diff
@@ -591,6 +709,59 @@ int main() {
...
+		s = 0
+		i += 1
+		while i > 0:
+			s += self.tree[i]
+			i -= i & (-i)
+		return s
+	
+	def get_sum(self, l, r):
+		return self.prefix_sum(r) - self.prefix_sum(l - 1)
+# --- END FENWICK TREE ---
```
Comment (SansPapyrus683) [inline content/4_Gold/PURS.mdx:743]:
```suggestion
```

File: content/4_Gold/PURS.mdx [hunk 7]
```diff
@@ -591,6 +709,59 @@ int main() {
...
 ```
+</CPPSection>
+<PySection>
+```python
+import sys
+input = sys.stdin.readline
+inf = float('inf')
+
+# --- BEGIN FENWICK TREE ---
+class FenwickTree:
```
Comment (SansPapyrus683) [inline content/4_Gold/PURS.mdx:720]: rename to `BIT`

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

**Comment by SansPapyrus683** [comment]
use tabs to indent

File: content/4_Gold/PURS.mdx [hunk 8]
```diff
@@ -314,14 +314,68 @@ public class MinSegmentTree {
 ```
 
 </JavaSection>
+
+<PySection>
+```python
+import sys
+inf = float('inf')
```
Comment (SansPapyrus683) [inline content/4_Gold/PURS.mdx:321]: just use `float('inf')` directly

File: content/4_Gold/PURS.mdx [hunk 9]
```diff
@@ -314,14 +314,68 @@ public class MinSegmentTree {
...
+input = sys.stdin.readline
```
Comment (SansPapyrus683) [inline content/4_Gold/PURS.mdx:322]: same with this function

File: content/4_Gold/PURS.mdx [hunk 10]
```diff
@@ -314,14 +314,68 @@ public class MinSegmentTree {
...
+		self.arr = arr
+		self.n = n
+		self.tree = [0] * (2 * n)
+
+		for ii in range(n):
+			self.tree[n + ii] = arr[ii]
+		for ii in reversed(range(n)):
+			self.tree[ii] = min(self.tree[ii << 1], self.tree[(ii << 1) | 1])
+
+	def update(self, i: int, v: int) -> None:
```
Comment (SansPapyrus683) [inline content/4_Gold/PURS.mdx:335]: add doc comments for `update` and `query` like in the module: https://usaco.guide/gold/PURS
also rename them to better describe what they do

File: content/4_Gold/PURS.mdx [hunk 11]
```diff
@@ -591,6 +694,52 @@ int main() {
...
+array = list(map(int, input().split()))
+array_bit = BIT(array, n)
+
+for _ in range(q):
+	q_type, k, u = map(int, input().split())
+	if q_type == 1:
+		array_bit.add_val(k - 1, u - array[k - 1])
+		array[k - 1] = u  # Make sure to update value after changing!
+	elif q_type == 2:
+		print(array_bit.get_sum(min(k, u)-1, max(k, u)-1))
```
Comment (SansPapyrus683) [inline content/4_Gold/PURS.mdx:738]: spaces between ops

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/4_Gold/PURS.mdx [hunk 12]
```diff
@@ -314,14 +314,69 @@ public class MinSegmentTree {
 ```
 
 </JavaSection>
+
+<PySection>
+```python
+import sys
+
+# A data structure that can answer point update & range minimum queries.
+class SegmentTree:
```
Comment (SansPapyrus683) [inline content/4_Gold/PURS.mdx:323]:
```suggestion
class SegmentTree:
	"""A data structure that can answer point update & range minimum queries."""
```
similar things for the other comments

File: content/4_Gold/PURS.mdx [hunk 13]
```diff
@@ -314,14 +314,69 @@ public class MinSegmentTree {
...
+		self.n = n
+		self.tree = [0] * (2 * n)
+
+		for ii in range(n):
+			self.tree[n + ii] = arr[ii]
+		for ii in reversed(range(n)):
+			self.tree[ii] = min(self.tree[ii << 1], self.tree[(ii << 1) | 1])
+
+	"""Adds v to the element at index i."""
+	def add(self, i: int, v: int) -> None:
```
Comment (SansPapyrus683) [inline content/4_Gold/PURS.mdx:335]: swap the order
```suggestion
	def add(self, i: int, v: int) -> None:
	"""Adds v to the element at index i."""
```

**Review by SansPapyrus683 — APPROVED** [review]
@envyaims

**Review by envyaims — APPROVED** [review]
el piethon

**Comment by thatpythonguys** [comment]
Would you be able to add my name to "Contributors" at the top? It's Pramana Saldin

**Comment by SansPapyrus683** [comment]
can you make another pr for that

**Comment by thatpythonguys** [comment]
Sure

**Pull request merged**